ReSharper.FSharp/src/FSharp/FSharp.Psi.Intentions/src/Intentions/FunctionAnnotationAction.fs

```diff
@@ -208,28 +226,54 @@ module SpecifyTypes =
 
         | _ -> ()
 
+module SpecifyTypesActionHelper =
+    open SpecifyTypes
+    let executePsiTransaction (node: ITreeNode) (availability: Availability) =
+        use writeCookie = WriteLockCookie.Create(node.IsPhysical())
+        use disableFormatter = new DisableCodeFormatter()
+        specifyTypes node availability
+
 
-[<ContextAction(Name = "AnnotateFunction", GroupType = typeof<FSharpContextActions>,
-                Description = "Annotate function with parameter types and return type")>]
-type FunctionAnnotationAction(dataProvider: FSharpContextActionDataProvider) =
+[<AbstractClass>]
+type AnnotationActionBase<'a when 'a: not struct and 'a :> ITreeNode>(dataProvider: FSharpContextActionDataProvider) =
     inherit FSharpContextActionBase(dataProvider)
 
-    override x.Text = "Add type annotations"
+    abstract member IsAvailable: 'a -> bool
+    member x.ContextNode = dataProvider.GetSelectedElement<'a>()
 
-    override x.IsAvailable _ =
-        let binding = dataProvider.GetSelectedElement<IBinding>()
-        if isNull binding then false else
-        if not (isAtBindingKeywordOrReferencePattern dataProvider binding) then false else
-        SpecifyTypes.getAvailability binding |> _.IsAvailable
+    override x.IsAvailable(_: IUserDataHolder) =
+        let node = x.ContextNode
+
+        isNotNull node && isValid (node :> ITreeNode) &&
```

Comment:
Are both `isValid` and `isNotNull` checks needed here? Doesn't the first one include the second? Also why is the cast needed?